### ISSUE TITLE
Spike - Research librdkafka / Sarama Capabilities

### ISSUE THREAD

**Issue by travis-minke-sap** [kyma-incubator/knative-kafka#114]
It has been a long time since we performed the original comparison between the librdkafka/confluent and sarama libraries.  We have not yet upgraded librdkafka and before we invest in doing so we want to re-evaluate whether it is still the best option.  Specifically we need the resulting implementation to be the highest throughput solution with elastic scaling and per-partition ordering.  There are concerns in knative about this library choice and we need solid information to back our selection (whatever it may be).

### Notes

Previously Sarama was split into two libraries (sarama and sarama-cluster) and they've now been integrated.  Sarama also had limitations regarding ordering that discounted it from consideration.  They do make it easy to read multiple partitions concurrently in the same consumer.  They don't yet support manual offset committing (though there are a number of issues open asking about it).

The librdkafka/confluent library is difficult to work with from a build/packaging perspective and there are concerns that crossing the Go/C barrier might be slow/inefficient.  Further, to move forward we'd need to support Go Modules which is what the confluent client is now based on.  It is also unclear whether they will support reading from specific partitions, but someone has created a PR to support reading from a specific partition https://github.com/confluentinc/confluent-kafka-go/pull/456 for it.

**Comment by travis-minke-sap** [comment]
CloudEvents sdk-go V2 has protocol support for sarama but not confluent...
https://github.com/cloudevents/sdk-go/tree/master/v2/protocol

**Comment by slinkydeveloper** [comment]
If it can help you, i have an "incomplete" benchmark of the different kafka clients out there: https://github.com/slinkydeveloper/kafka-go-clients-benchmarks
You can use it as basis for benchmarking, but you'll need to hack the librdkafka code (i have to upload it)

Tl;dr librdkafka performs well on producer side, but consuming is slow because it's one polling entrypoint for consumer, not one per partition as consumer group by sarama. There is an open issue about that: https://github.com/confluentinc/confluent-kafka-go/issues/440

**Comment by pbochynski** [comment]
Closing all issues as project is going to be archived. If it is still valid create a new issue in https://github.com/knative-sandbox/eventing-kafka
